### ISSUE TITLE
Performance: Preloading components in "tree" hierarchy could help with perf

### ISSUE THREAD

**Issue by leeviana** [microsoft/FluidFramework#7256]
## Feature
From the perf team: One performance problem we have today is that we don't discover the subcomponents we need to render on the critical path until it is far too late. For example, consider the case of rendering a component that happens to host another. Today you have to sequentially load those components and their code/factories.

Now that the framework has a concept of a hierarchy with garbage collection, some of this loading could happen in parallel, which may result in drastic perf improvements.

**Other things noted for consideration:**
* Today, we load a snapshot that contains all the component blobs in it. If this assumption changes in the future, such that some component blobs are fetched lazily, we would like to similarly fetch these subcomponent blobs early.
* If we were to modify the Fluid load contract to no longer return all blobs on load for a container, but rather just the subset of "critical" blobs for the initial render, how can we make sure that subcomponents get treated as critical blobs, and not just the blobs corresponding to the top level component.
* This problem could also exist for outies... so we could keep that in mind, though I don't think any container hierarchy solutions would help here.

**Comment by leeviana** [comment]
Comment to remove stale tag.

**Comment by tylerbutler** [comment]
@microsoft/fluid-cr-loader Any thoughts on this request?

**Comment by leeviana** [comment]
not stale

**Comment by vladsud** [comment]
@leeviana, as it was originally opened, this issue seems about nothing. I.e. I read it - "if runtime behavior changes, please account for X". Given that all content is cached in memory today, any pre-loading is basically noop. We should not have issues like that :)

That said, we are very likely to move toward more shredded storage / loading, as described in https://dev.azure.com/fluidframework/internal/_queries/edit/2319/?triage=true.

But the purpose / implementation (including policies) will be about
- maintaining single round-trip for most small documents, while allowing larger docs (including bigger than memory) to work correctly.
- System to work correctly in presence of SPO cut-off limits for snapshots and reducing these limits from today's 200Mb to much smaller sizes (I'd guess in the vicinity of 8-16Mb).

So while it fundamentally does not change much, one thing to watch for is large docs, and majority of them - containing garbage. I.e. we need GC enabled the sooner the better.

As for runtime having more means to do pre-fetching, we are looking for unification of channels (i.e. supporting nested data stores) as a way to express such dependencies more explicitly and allowing combination of runtime & storage to optimize scenarios like you mention (by loading full sub-trees in one go, if they are small).

At the moment, I do not see a value in keeping this task around, but feel free to comment in above mentioned ADO item.

**Comment by vladsud** [comment]
FF does not do any pre-loading - everything is delay-loaded. I do not think this behavior should be baked into runtime - this may result in whole document being loading, depending on scenario, and may even result in out-of-memory condition for documents larger than memory.

It should be done as a layer on top of FF. Runtime does not have that much any advantage here really - someone need to load DDS and read all content. If by doing so we learn that it's too slow, and somehow runtime has an edge to cut that time substantially, then we should look into adding more capabilities to runtime.

Basically, we should be data driven, and keep runtime layers as slip and simple as possible.
